packages/core/src/actors.ts

```diff
@@ -244,68 +251,56 @@ export function fromObservable<T, TEvent extends EventObject>(
   const completeEventType = '$$xstate.complete';
 
   // TODO: add event types
-  const behavior: Behavior<
-    any,
-    T | undefined,
-    {
-      subscription: Subscription | undefined;
-      observable: Subscribable<T> | undefined;
-      canceled: boolean;
-      status: 'active' | 'done' | 'error';
-      data: T | undefined;
-    }
-  > = {
-    transition: (state, event, { self, id }) => {
+  const behavior: Behavior<any, T | undefined, ObservableInternalState<T>> = {
+    transition: (state, event, { self, id, defer }) => {
       const _event = toSCXMLEvent(event);
+
       if (state.canceled) {
         return state;
       }
 
       switch (_event.name) {
         case nextEventType:
-          self._parent?.send(
-            toSCXMLEvent(
-              {
-                type: `xstate.snapshot.${id}`,
-                data: _event.data.data
-              },
-              { origin: self }
-            )
-          );
-          return { ...state, data: event.data.data };
+          state.data = event.data.data;
+          // match the exact timing of events sent by machines
+          // send actions are not executed immediately
+          defer(() => {
```

Comment:
Technically this isn't needed right now. Things work OK with this PR without this `defer` - that's because in this PR we start mutating `state` directly. If the `state` would be updated in an immutable way (like before this PR) then this `defer` would be very much needed to make the added test pass:
https://github.com/statelyai/xstate/pull/3740/files#diff-8ecfc4c99852fea354bfd5cf3c0a171ff48556ef07a9fd9d588aceaae25969f5R440

I still think though that it's better to have this `defer` here in place - to match the timing of things here as close to the timing within machines